data/marshaller_test.go

```diff
@@ -7,73 +7,99 @@ import (
+		ExpectedContentType: contentTypeApplicationXML,
+		ExpectedOutput:      `<Data><A>1</A><B>foo</B><C>foo</C><C>bar</C><C>baz</C></Data>`,
+	},
+	{
+		Data: Data{
+			1,
+			"foo",
+			[]string{"foo", "bar", "baz"},
 		},
-	}
-
-	rec := httptest.NewRecorder()
-
-	Marshal(rec, request, subject)
-
-	response := rec.Result()
+		Accept:              contentTypeTextXML,
+		ExpectedContentType: contentTypeApplicationXML,
+		ExpectedOutput:      `<Data><A>1</A><B>foo</B><C>foo</C><C>bar</C><C>baz</C></Data>`,
+	},
+	{
+		Data: Data{
+			1,
+			"foo",
+			[]string{"foo", "bar", "baz"},
+		},
+		Accept:              contentTypeJSON,
+		ExpectedContentType: contentTypeJSON,
+		ExpectedOutput:      `{"A":1,"B":"foo","C":["foo","bar","baz"]}`,
+	},
+	{
+		Data:                []string{"foo", "bar", "baz"},
+		Accept:              contentTypeApplicationXML,
+		ExpectedContentType: contentTypeApplicationXML,
+		ExpectedOutput:      `<result length="3"><Results>foo</Results><Results>bar</Results><Results>baz</Results></result>`,
+	},
+}
 
-	if rec.Code != http.StatusOK {
-		t.Error("Marshaller failed with an unknown error")
+func TestMarshal(t *testing.T) {
+
+	for _, tc := range testCases {
+		request := httptest.NewRequest("GET", "http://missy.local/marshal", nil)
+		request.Header.Set("Accept", tc.Accept)
+		response := httptest.NewRecorder()
+		Marshal(response, request, tc.Data)
+		if ct := response.Header().Get(httpHeaderContentType); ct != tc.ExpectedContentType {
+			t.Logf("Accept header has the wrong content type. Expected %s, actual %s", tc.ExpectedContentType, ct)
+			t.Fail()
+		}
+		if response.Code != http.StatusOK {
+			t.Log("Marshaller returned not StatusCode 200 OK")
+			t.Fail()
+		}
+		bodyBytes, _ := ioutil.ReadAll(response.Body)
+		if strBody := string(bodyBytes); strBody != tc.ExpectedOutput {
+			t.Errorf("Body does not match expected output. Expected %s, actual %s", tc.ExpectedOutput, strBody)
+		}
 	}
+}
 
-	if ct := response.Header.Get(httpHeaderContentType); ct != contentTypeApplicationXML {
-		t.Errorf("Accept header has the wrong content type. Expected %s, actual %s", contentTypeApplicationXML, ct)
-	}
-	bodyBytes, _ := ioutil.ReadAll(rec.Body)
-	if strBody := string(bodyBytes); strBody != expectedXML {
-		t.Errorf("Body does not match expected output. Expected %s, actual %s", expectedXML, strBody)
+func TestMarshalWithCode(t *testing.T) {
+	for _, tc := range testCases {
+		request := httptest.NewRequest("GET", "http://missy.local/marshal", nil)
+		request.Header.Set("Accept", tc.Accept)
+		response := httptest.NewRecorder()
+		MarshalWithCode(response, request, tc.Data, http.StatusTeapot)
+		if ct := response.Header().Get(httpHeaderContentType); ct != tc.ExpectedContentType {
+			t.Logf("Accept header has the wrong content type. Expected %s, actual %s", tc.ExpectedContentType, ct)
+			t.Fail()
+		}
+		if response.Code != http.StatusTeapot {
+			t.Log("Marshaller returned not StatusCode 418 I am a Teapot")
```

Comment:
Actually it's defined as `I'm a teapot` in spec.

Reply:
nit picker alert

Reply:
You need to live with that fact :D